### PR TITLE
Orderline

### PR DESCRIPTION
**Pull request by Oleksandr-Radovets**

### CODE REVIEW

File: src/main/java/com/example/securitynew/mapper/OrderMapper.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface OrderMapper {
+    OrderResponseDto toDto(Order order);
+
+    OrderResponseStatusDto toDtoStatus(Order order);
+
+    OrderResponseOrderItemDto toDtoOrderItem(OrderItem orderitem);
+
```
Comment (MentorMA) [inline src/main/java/com/example/securitynew/mapper/OrderMapper.java:16]:
```suggestion
```

File: src/main/java/com/example/securitynew/model/CartItem.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    private ShoppingCart shoppingCart;
+    @ManyToOne
+    @JoinTable(name = "cartItem_books",
+            joinColumns = @JoinColumn(name = "cart_item_id"),
+            inverseJoinColumns = @JoinColumn(name = "book_id"))
+    private Book book;
+    private int quantity;
+    private boolean isDeleted;
+
+
```
Comment (MentorMA) [inline src/main/java/com/example/securitynew/model/CartItem.java]:
```suggestion
```

File: src/main/java/com/example/securitynew/repository/OrderItemRepository.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.example.securitynew.repository;
+
+import com.example.securitynew.model.OrderItem;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {
+
+
```
Comment (MentorMA) [inline src/main/java/com/example/securitynew/repository/OrderItemRepository.java]:
```suggestion
```

File: src/main/java/com/example/securitynew/repository/OrderRepository.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.example.securitynew.repository;
+
+import com.example.securitynew.model.Order;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface OrderRepository extends JpaRepository<Order, Long> {
+}
```
Comment (MentorMA) [inline src/main/java/com/example/securitynew/repository/OrderRepository.java]: Keep same style of interface

File: src/main/java/com/example/securitynew/service/CartItemService.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+import com.example.securitynew.model.CartItem;
+
+public interface CartItemService {
+
+    void deleteCartItem(Long id);
+
+    CartItem saveCartItem(CartItem cartItem);
+
+    CartItem cartItemFindById(Long id);
+
```
Comment (MentorMA) [inline src/main/java/com/example/securitynew/service/CartItemService.java]:
```suggestion
```

**Review by MentorMA — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/securitynew/dto/order/OrderRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class OrderRequestDto {
+    private Long id;
+    private String shippingAddress;
+    private String status;
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/dto/order/OrderRequestDto.java]:
```suggestion
public class OrderRequestDto {
    private String shippingAddress;
```

File: src/main/java/com/example/securitynew/dto/order/OrderResponseOrderItemDto.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package com.example.securitynew.dto.order;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class OrderResponseOrderItemDto {
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/dto/order/OrderResponseOrderItemDto.java]: fix class name

File: src/main/java/com/example/securitynew/dto/order/OrderResponseStatusDto.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.example.securitynew.dto.order;
+
+import com.example.securitynew.model.Status;
+
+public class OrderResponseStatusDto {
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/dto/order/OrderResponseStatusDto.java]: fix class name

File: src/main/java/com/example/securitynew/mapper/OrderItemMapper.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package com.example.securitynew.mapper;
+
+import com.example.securitynew.config.MapperConfig;
+import com.example.securitynew.model.CartItem;
+import com.example.securitynew.model.OrderItem;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface OrderItemMapper {
+    OrderItem orderItemMapperList(CartItem cartItem);
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/mapper/OrderItemMapper.java]: incorrect method name - it should have verb or "to..."

File: src/main/java/com/example/securitynew/controller/CategoryController.java [hunk 10]
```diff
@@ -16,7 +17,6 @@
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
-import java.util.List;
 
 @RestController
 @RequiredArgsConstructor
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/CategoryController.java:22]: Using @RequiredArgsConstractor you want all components to be autowired via constructor, to do so you need to put final key word (https://www.baeldung.com/java-lombok-constructor-annotation)

File: src/main/java/com/example/securitynew/controller/OrderController.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PatchMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequestMapping("/api/orders")
+@RequiredArgsConstructor
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/OrderController.java:21]: `final` for components

File: src/main/java/com/example/securitynew/controller/OrderController.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping("/order")
+    public OrderResponseDto saveOrder(
+            @RequestBody OrderRequestDto orderRequestDto) {
+        return orderService.createOrder(orderRequestDto.getShippingAddress());
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("/all")
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/OrderController.java]: `all` change with `orders`

File: src/main/java/com/example/securitynew/controller/OrderController.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    public List<OrderResponseDto> allOrders() {
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/OrderController.java]: Method name requires some verb to understand what method does. Change it to `getOrders`

File: src/main/java/com/example/securitynew/controller/CategoryController.java [hunk 14]
```diff
@@ -27,7 +27,8 @@ public class CategoryController {
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/CategoryController.java:27]: @RequestMapping("/api/categories")

File: src/main/java/com/example/securitynew/controller/CategoryController.java [hunk 15]
```diff
@@ -27,7 +27,8 @@ public class CategoryController {
 
     @PreAuthorize("has_Role('ROLE_ADMIN')")
     @PostMapping("/create")
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/CategoryController.java]: According to rest `create` is wrong. Just POST: /api/categories

File: src/main/java/com/example/securitynew/controller/CategoryController.java [hunk 16]
```diff
@@ -48,20 +49,20 @@ public CategoryResponseDto getCategoryById(@PathVariable Long id) {
     @PreAuthorize("has_Role('ROLE_ADMIN')")
     @PutMapping("/update/{id}")
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/CategoryController.java]: PUT: /api/categories/{id}

File: src/main/java/com/example/securitynew/controller/CategoryController.java [hunk 17]
```diff
@@ -48,20 +49,20 @@ public CategoryResponseDto getCategoryById(@PathVariable Long id) {
...
     @DeleteMapping("/deleteCategory/{id}")
-    public void deleteCategory(@PathVariable Long id){
-         categoryService.deleteById(id);
+    public void deleteCategory(@PathVariable Long id) {
+        categoryService.deleteById(id);
     }
 
     @Operation(summary = "get Book by Category id", description = "get Book by Category id")
     @PreAuthorize("has_Role('ROLE_USER')")
     @GetMapping("/allBookByCategory/{id}")
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/CategoryController.java]: check all requirements for endpoints

File: src/main/java/com/example/securitynew/controller/OrderController.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+    @PatchMapping("/update/{orderId}")
+    public OrderResponseStatusDto updateStatusOrderItem(@PathVariable Long orderId,
+                                                        @RequestBody
+                                                        OrderRequestDto orderRequestDto) {
+        return orderService.update(orderId, orderRequestDto.getStatus());
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("/allOrderItem/{orderId}")
+    public List<OrderResponseOrderItemDto> getAllOrderItem(@PathVariable Long orderId) {
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/OrderController.java]:
```suggestion
    public List<OrderResponseOrderItemDto> getAllOrderItems(@PathVariable Long orderId) {
```

File: src/main/java/com/example/securitynew/controller/OrderController.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+        return orderService.getAllOrderItemByOrder(orderId);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("/api/orders/{orderId}/items/{itemId}")
+    public OrderResponseOrderItemDto getOrderItemByIdOrder(@PathVariable Long orderId,
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/OrderController.java]:
```suggestion
    public OrderResponseOrderItemDto getOrderItem(@PathVariable Long orderId,
```

File: src/main/java/com/example/securitynew/controller/OrderController.java [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("/all")
+    public List<OrderResponseDto> allOrders() {
+        return orderService.allOrdersUser();
+    }
+
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @PatchMapping("/update/{orderId}")
+    public OrderResponseStatusDto updateStatusOrderItem(@PathVariable Long orderId,
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/OrderController.java]:
```suggestion
    public OrderResponseStatusDto updateOrderItemStatus(@PathVariable Long orderId,
```

File: src/main/java/com/example/securitynew/controller/ShoppingCartController.java [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/cart")
+public class ShoppingCartController {
+    private ShoppingCartService shoppingCartService;
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("/cart/{id}")
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/ShoppingCartController.java:24]: Redundant `/cart`. You have @RequestMapping("/api/cart").

File: src/main/java/com/example/securitynew/controller/OrderController.java [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("/allOrderItem/{orderId}")
+    public List<OrderResponseOrderItemDto> getAllOrderItem(@PathVariable Long orderId) {
+        return orderService.getAllOrderItemByOrder(orderId);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("/api/orders/{orderId}/items/{itemId}")
+    public OrderResponseOrderItemDto getOrderItemByIdOrder(@PathVariable Long orderId,
+                                                           @PathVariable Long orderItemId) {
```
Comment (artemZhuravskyi) [inline src/main/java/com/example/securitynew/controller/OrderController.java]:
```suggestion
                                                           @PathVariable Long itemId) {
```

**Review by artemZhuravskyi — CHANGES_REQUESTED** [review]
1) Please, test your application before creating Pull Request.
2) Check endpoint namings.
3) Don't use so many DTOs for one object.

**Comment by Oleksandr-Radovets** [comment]
my tasks are only Order

File: src/main/java/com/example/securitynew/controller/OrderController.java [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("/orders")
+    public List<OrderResponseDto> getOrders() {
+        return orderService.allOrdersUser();
+    }
+
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @PatchMapping("/update/{orderId}")
```
Comment (Rommelua) [inline src/main/java/com/example/securitynew/controller/OrderController.java]: wrong REST mapping
consider using class-level and method-level mappings

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/securitynew/controller/CategoryController.java (hunk 16 above)
Comment (koval-kn) [reply src/main/java/com/example/securitynew/controller/CategoryController.java]: not fixed

File: src/main/java/com/example/securitynew/controller/CategoryController.java (hunk 15 above)
Comment (koval-kn) [reply src/main/java/com/example/securitynew/controller/CategoryController.java]: not fixed

File: src/main/java/com/example/securitynew/controller/CategoryController.java (hunk 14 above)
Comment (koval-kn) [reply src/main/java/com/example/securitynew/controller/CategoryController.java:27]: not fixed

File: src/main/java/com/example/securitynew/controller/CategoryController.java (hunk 17 above)
Comment (koval-kn) [reply src/main/java/com/example/securitynew/controller/CategoryController.java]: not fixed

File: src/main/java/com/example/securitynew/controller/ShoppingCartController.java (hunk 21 above)
Comment (koval-kn) [reply src/main/java/com/example/securitynew/controller/ShoppingCartController.java:24]: not fixed

**Review by koval-kn — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/securitynew/controller/CategoryController.java (hunk 14 above)
Comment (koval-kn) [reply src/main/java/com/example/securitynew/controller/CategoryController.java:27]: not fixed

File: src/main/java/com/example/securitynew/controller/CategoryController.java [hunk 24]
```diff
@@ -16,52 +17,52 @@
...
 import org.springframework.web.bind.annotation.RestController;
-import java.util.List;
 
 @RestController
 @RequiredArgsConstructor
-@RequestMapping("/api/category")
+@RequestMapping("/api")
 public class CategoryController {
     private CategoryService categoryService;
     private BookService bookService;
```
Comment (koval-kn) [inline src/main/java/com/example/securitynew/controller/CategoryController.java]:
```suggestion
    private final CategoryService categoryService;
    private final BookService bookService;

```

**Review by koval-kn — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/securitynew/controller/CategoryController.java (hunk 14 above)
Comment (Vasul-Pankiv) [reply src/main/java/com/example/securitynew/controller/CategoryController.java:27]: are you kidding?? Please, FIX requestMapping
image

File: src/main/java/com/example/securitynew/controller/ShoppingCartController.java (hunk 21 above)
Comment (Vasul-Pankiv) [reply src/main/java/com/example/securitynew/controller/ShoppingCartController.java:24]: same here, not fixed

**Review by Vasul-Pankiv — CHANGES_REQUESTED** [review]

**Comment by Oleksandr-Radovets** [comment]
мені здається що ви не розумієте як в завданні написано я індентично так само здроби , складається враження що вине розумієте завдань по яким прохлдим

**Review by koval-kn — APPROVED** [review]
congratz :)

**Pull request merged**